api/api.go

```diff
@@ -31,109 +30,50 @@ var (
 type API struct {
 	handler    http.Handler
 	db         *gorm.DB
-	paypal     *paypalsdk.Client
-	config     *conf.Configuration
-	mailer     mailer.Mailer
+	config     *conf.GlobalConfiguration
 	httpClient *http.Client
 	log        *logrus.Entry
-	assets     assetstores.Store
 	version    string
 }
 
-type JWTClaims struct {
-	ID           string                 `json:"id"`
-	Email        string                 `json:"email"`
-	AppMetaData  map[string]interface{} `json:"app_metadata"`
-	UserMetaData map[string]interface{} `json:"user_metadata"`
-	*jwt.StandardClaims
+// ListenAndServe starts the REST API.
+func (a *API) ListenAndServe(hostAndPort string) error {
+	return http.ListenAndServe(hostAndPort, a.handler)
 }
 
-func (a *API) withToken(ctx context.Context, w http.ResponseWriter, r *http.Request) context.Context {
-	log := getLogger(ctx)
-	config := getConfig(ctx)
-	authHeader := r.Header.Get("Authorization")
-	if authHeader == "" {
-		log.Info("Making unauthenticated request")
-		return ctx
-	}
+// NewAPI instantiates a new REST API using the default version.
+func NewAPI(globalConfig *conf.GlobalConfiguration, config *conf.Configuration, db *gorm.DB) *API {
+	return NewSingleTenantAPIWithVersion(globalConfig, config, db, defaultVersion)
+}
 
-	matches := bearerRegexp.FindStringSubmatch(authHeader)
-	if len(matches) != 2 {
-		log.Info("Invalid auth header format: " + authHeader)
-		unauthorizedError(w, "Bad authentication header")
-		return nil
-	}
+// NewMultiTenantAPI creates a new REST API to serve multiple tenants.
+func NewMultiTenantAPI(globalConfig *conf.GlobalConfiguration, db *gorm.DB) *API {
+	return NewAPIWithVersion(context.Background(), globalConfig, db, defaultVersion)
+}
 
-	token, err := jwt.ParseWithClaims(matches[1], &JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
-		if token.Header["alg"] != jwt.SigningMethodHS256.Name {
-			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
-		}
-		return []byte(config.JWT.Secret), nil
-	})
+// NewSingleTenantAPIWithVersion creates a single-tenant version of the REST API
+func NewSingleTenantAPIWithVersion(globalConfig *conf.GlobalConfiguration, config *conf.Configuration, db *gorm.DB, version string) *API {
+	ctx := context.Background()
+	ctx, err := withTenantConfig(ctx, config)
 	if err != nil {
-		log.Infof("Invalid token: %v", err)
-		unauthorizedError(w, "Invalid token")
-		return nil
-	}
-
-	claims := token.Claims.(*JWTClaims)
-	if claims.StandardClaims.ExpiresAt < time.Now().Unix() {
-		msg := fmt.Sprintf("Token expired at %v", time.Unix(claims.StandardClaims.ExpiresAt, 0))
-		log.Info(msg)
-		unauthorizedError(w, msg)
-		return nil
-	}
-
-	isAdmin := false
-	roles, ok := claims.AppMetaData["roles"]
-	if ok {
-		roleStrings, _ := roles.([]interface{})
-		for _, data := range roleStrings {
-			role, _ := data.(string)
-			if role == config.JWT.AdminGroupName {
-				isAdmin = true
-				break
-			}
-		}
+		logrus.Fatalf("%+v", err)
```

Comment:
The logger is passed to the API [here](https://github.com/netlify/gocommerce/blob/345a7cd5f13fe5d51ab1107bf389fffd291b75a8/api/api.go#L68).